pkg/restore/restore.go

```diff
@@ -708,251 +727,335 @@ func (ctx *context) restoreResource(resource, namespace, resourcePath string) (a
+			delete(pvc.Annotations, "pv.kubernetes.io/bound-by-controller")
+
+			res, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pvc)
+			if err != nil {
 				addToResult(&errs, namespace, err)
-				continue
+				return warnings, errs
 			}
+			obj.Object = res
+		}
+	}
+
+	// clear out non-core metadata fields & status
+	if obj, err = resetMetadataAndStatus(obj); err != nil {
+		addToResult(&errs, namespace, err)
+		return warnings, errs
+	}
 
-			if pvc.Spec.VolumeName != "" && ctx.pvsToProvision.Has(pvc.Spec.VolumeName) {
-				ctx.log.Infof("Resetting PersistentVolumeClaim %s/%s for dynamic provisioning because its PV %v has a reclaim policy of Delete", namespace, name, pvc.Spec.VolumeName)
+	for _, action := range ctx.getApplicableActions(groupResource, namespace) {
+		if !action.selector.Matches(labels.Set(obj.GetLabels())) {
+			return warnings, errs
+		}
 
-				pvc.Spec.VolumeName = ""
-				delete(pvc.Annotations, "pv.kubernetes.io/bind-completed")
-				delete(pvc.Annotations, "pv.kubernetes.io/bound-by-controller")
+		ctx.log.Infof("Executing item action for %v", &groupResource)
 
-				res, err := runtime.DefaultUnstructuredConverter.ToUnstructured(pvc)
-				if err != nil {
-					addToResult(&errs, namespace, err)
-					continue
-				}
-				obj.Object = res
-			}
+		executeOutput, err := action.Execute(&velero.RestoreItemActionExecuteInput{
+			Item:           obj,
+			ItemFromBackup: itemFromBackup,
+			Restore:        ctx.restore,
+		})
+		if executeOutput.Warning != nil {
+			addToResult(&warnings, namespace, fmt.Errorf("warning preparing %s: %v", resourceID, executeOutput.Warning))
+		}
+		if err != nil {
+			addToResult(&errs, namespace, fmt.Errorf("error preparing %s: %v", resourceID, err))
+			return warnings, errs
 		}
 
-		// clear out non-core metadata fields & status
-		if obj, err = resetMetadataAndStatus(obj); err != nil {
-			addToResult(&errs, namespace, err)
-			continue
+		unstructuredObj, ok := executeOutput.UpdatedItem.(*unstructured.Unstructured)
+		if !ok {
+			addToResult(&errs, namespace, fmt.Errorf("%s: unexpected type %T", resourceID, executeOutput.UpdatedItem))
+			return warnings, errs
 		}
 
-		for _, action := range applicableActions {
-			if !action.selector.Matches(labels.Set(obj.GetLabels())) {
-				continue
+		obj = unstructuredObj
+
+		for _, additionalItem := range executeOutput.AdditionalItems {
+			var itemPath string
+			if additionalItem.Namespace == "" {
+				itemPath = filepath.Join(ctx.restoreDir, api.ResourcesDir, additionalItem.GroupResource.String(), api.ClusterScopedDir, additionalItem.Name+".json")
+			} else {
+				itemPath = filepath.Join(ctx.restoreDir, api.ResourcesDir, additionalItem.GroupResource.String(), api.NamespaceScopedDir, additionalItem.Namespace, additionalItem.Name+".json")
 			}
 
-			ctx.log.Infof("Executing item action for %v", &groupResource)
+			if _, err := ctx.fileSystem.Stat(itemPath); err != nil {
+				ctx.log.WithError(err).WithFields(logrus.Fields{
```

Comment:
TODO is logging this as a warning the right thing to do? might want to add it to the restore results?

Reply:
Since we already are returning warnings, then it makes more sense to return all warnings instead of returning some and login some.